### DISCUSSION TITLE
What's the simplest way to leave a gap in the middle of the QR code?

### DISCUSSION THREAD

**Discussion by boli** [chillerlan/php-qrcode#133] [Q&A]
No image. Just a square blank space...

**Comment by codemasher** [comment]
Hey, in case you're using the current `dev-main` (v5) branch, you can just set `QROptions::$addLogoSpace` to `true` and modify the size of the space with `QROptions::$logoSpaceWidth` and `QROptions::$logoSpaceHeight` (relevant bit [over here](https://github.com/chillerlan/php-qrcode/blob/cde2af576b6b829b1a30e3042eba488c90d94478/src/QRCode.php#L172-L180)).

For the released v4.3.x you could do something like this (see also [`QRImageWithLogo`](https://github.com/chillerlan/php-qrcode/blob/v4.3.x/examples/QRImageWithLogo.php) and [`imageWithLogo`](https://github.com/chillerlan/php-qrcode/blob/v4.3.x/examples/imageWithLogo.php)):

```php
// extend the options class to add size parameters
class LogoSpaceOptions extends QROptions{
	// size in QR modules
	protected int $logoSpaceWidth;
	protected int $logoSpaceHeight;
}

// this would work with any output module, GD is just for example
class QRImageWithLogo extends QRImage{

	public function dump(string $file = null){

		// add logo space before processing the output
		// actually, you can freely modify the matrix at this point via QRMatrix::set()
		$this->matrix->setLogoSpace(
			$this->options->logoSpaceWidth,
			$this->options->logoSpaceHeight
		);

		return parent::dump($file);
	}

}
```

**Reply by brewedbyanne** [reply]
I have a question on this. If I set `QROptions::$addLogoSpace` to true, is it just ignoring any bits on the matrix that would be filled in dark and relying on the error correction to fill those in?  And if so, is there a relatively 'safe' threshold i should stay below for the width/height?

**Reply by codemasher** [reply]
Hey, 

> If I set QROptions::$addLogoSpace to true, is it just ignoring any bits on the matrix that would be filled in dark and relying on the error correction to fill those in?

This is correct and the reason why it enforces ECC level H - currently there's an [internal maximum of 20% of the total modules](https://github.com/chillerlan/php-qrcode/blob/739579552d54d1b24a75415af805faba37b975fd/src/Data/QRMatrix.php#L556-L559) (excluding quiet zone). Technically i could map the modules around the logo space but this would be as much against the specification as adding logos is - most QR Codes with logos you see in the wild have just the logo slapped over the matrix.

**Reply by brewedbyanne** [reply]
Yeah, 100%. I was pretty sure that was the case but wanted to double check. Thank you for the quick reply (and the great library!)

**Reply by codemasher** [reply]
You're welcome!
